### PR TITLE
add pre-commit hook with clang-tidy and clang-format

### PR DESCRIPTION
**Pull request by alisa-vernigor**

### CODE REVIEW

**Review by SphericalPotatoInVacuum — CHANGES_REQUESTED** [review]
I suggest copying `.clang-format` from another branch (we can just delete that branch) and formatting all the code there already is, so new commits in other branches don't have to format everything.

File: .clang-tidy [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+  -readability-function-cognitive-complexity,
+  -bugprone-narrowing-conversions,
+  -bugprone-easily-swappable-parameters,
+  -bugprone-implicit-widening-of-multiplication-result,
+  -bugprone-unchecked-optional-access
+
+# Turn all the warnings from the checks above into errors.
+WarningsAsErrors: "*"
+
+HeaderFilterRegex: "(google/cloud/|generator/).*\\.h$"
```
Comment (SphericalPotatoInVacuum) [inline .clang-tidy:83]: I'm sure we need another header filter 😊

Comment (SphericalPotatoInVacuum) [inline .clang-tidy]: I'm sure we need another header filter :blush:

**Review by SphericalPotatoInVacuum — CHANGES_REQUESTED** [review]

**Comment by alisa-vernigor** [comment]
> I suggest copying `.clang-format` from another branch (we can just delete that branch) and formatting all the code there already is, so new commits in other branches don't have to format everything.

I suggest using "rebase main" in other branches.

**Review by SphericalPotatoInVacuum — APPROVED** [review]
LGTM

**Pull request merged**